service_container/injection_types.rst

```diff
@@ -100,6 +100,103 @@ working with optional dependencies. It is also more difficult to use in
 combination with class hierarchies: if a class uses constructor injection
 then extending it and overriding the constructor becomes problematic.
 
+Immutable-setter Injection
+--------------------------
+
+Another possible injection in to use a method which return ``static``,
+this approach allow you to make a service immutable::
+
+    // ...
+
+    class NewsletterManager
+    {
+        private $mailer;
+
+        /**
+         * @required
+         * @return static
+         */
+        public function withMailer(MailerInterface $mailer)
+        {
+            $new = clone $this;
+            $new->mailer = $mailer;
+
+            return $new;
+        }
+
+        // ...
+    }
+
+In order to use the full potential of this approach, you can define a third argument
```

Comment:
"full potential" is strange here - it doesn't work if you don't configure that way.
Maybe adapt the sentence I suggested above?

Reply:
Improved, to be checked 🙂